shims/spark32/src/main/scala/org/apache/spark/sql/execution/adaptive/GlutenCostEvaluator.scala

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.execution.SparkPlan
+
+case class GlutenCost(value: Long, planId: Int) extends Cost {
+  override def compare(that: Cost): Int = that match {
+    case GlutenCost(thatValue, thatId) =>
+      if (value < thatValue || (value == thatValue && planId != thatId)) -1
```

Comment:
The planId is increasing, but the parent node may be removed, causing the id to become smaller, such as `OptimizeOneRowPlan`, so use `!=`.

Reply:
This part is comparatively tricky and important in this PR so let's code it carefully. `planId != thatId` is still a condition relying on caller code.

Would you check if we can use something like this:

```scala
class GlutenCost(val eval: CostEvaluator, val plan: SparkPlan) extends Cost {
  override def compare(that: Cost): Int = that match {
    case that: GlutenCost if plan eq that.plan =>
      0
    case that: GlutenCost if plan == that.plan =>
      // Plans are identical. Considers the newer one as having lower cost.
      -(plan.id - that.plan.id)
    case that: GlutenCost =>
      // Plans are different. Use the delegated cost evaluator.
      assert(eval == that.eval)
      eval.evaluateCost(plan).compare(eval.evaluateCost(that.plan))
    case _ =>
      throw QueryExecutionErrors.cannotCompareCostWithTargetCostError(that.toString)
  }

  override def hashCode(): Int = throw new UnsupportedOperationException()

  override def equals(obj: Any): Boolean = obj match {
    case that: Cost => compare(that) == 0
    case _ => false
  }
}
```

Reply:
> ```scala
>       eval.evaluateCost(plan).compare(eval.evaluateCost(that.plan))
> ```


There seems to be a codependency issue here. 
How about this：

```
class GlutenCost(value: Long, plan: SparkPlan) extends SimpleCost(value) {
  override def compare(that: Cost): Int = that match {
    case that: GlutenCost if plan eq that.plan =>
      0
    case that: GlutenCost if plan == that.plan =>
      // Plans are identical. Considers the newer one as having lower cost.
      -(plan.id - that.plan.id)
    case _ =>
      // Plans are different. Use the default cost compare.
      super.compare(that)
  }

  override def hashCode(): Int = throw new UnsupportedOperationException()

  override def equals(obj: Any): Boolean = obj match {
    case that: Cost => compare(that) == 0
    case _ => false
  }
}

/** This [[CostEvaluator]] is to force use the new physical plan when cost is equal. */
case class GlutenCostEvaluator() extends CostEvaluator with SQLConfHelper {
  override def evaluateCost(plan: SparkPlan): Cost = {
    val forceOptimizeSkewedJoin = conf.getConf(SQLConf.ADAPTIVE_FORCE_OPTIMIZE_SKEWED_JOIN)
    val simpleCost = SimpleCostEvaluator(forceOptimizeSkewedJoin)
      .evaluateCost(plan)
      .asInstanceOf[SimpleCost]
    new GlutenCost(simpleCost.value, plan)
  }
}
```

Reply:
```
class GlutenCost(value: Long, plan: SparkPlan) extends SimpleCost(value)
```

I will not be into this approach. Class inherency (despite that it's a case-class inheritance) always make things more complicated.


> There seems to be a codependency issue here.

There should not be codependency. `val eval: CostEvaluator` is the one that used by Spark (SimpleCostEvaluator) rather than GlutenCostEvaluator. E.g.,


```scala
case class GlutenCostEvaluator() extends CostEvaluator with SQLConfHelper {
  override def evaluateCost(plan: SparkPlan): Cost = {
    val forceOptimizeSkewedJoin = conf.getConf(SQLConf.ADAPTIVE_FORCE_OPTIMIZE_SKEWED_JOIN)
    new GlutenCost(SimpleCostEvaluator(forceOptimizeSkewedJoin), plan)
  }
}
```

Reply:
SimpleCostEvaluator.evaluateCost return SimpleCost, not GlutenCost

Reply:
> SimpleCostEvaluator.evaluateCost return SimpleCost, not GlutenCost

It's intentional. Gluten only chooses the newest plan when plans are equal. Otherwise we just compare against SimpleCost.

Reply:
```
case class GlutenCostEvaluator() extends CostEvaluator with SQLConfHelper {
  override def evaluateCost(plan: SparkPlan): Cost = {
    val forceOptimizeSkewedJoin = conf.getConf(SQLConf.ADAPTIVE_FORCE_OPTIMIZE_SKEWED_JOIN)
    new GlutenCost(SimpleCostEvaluator(forceOptimizeSkewedJoin), plan)
  }
}
```